invokeai/backend/generator/base.py

```diff
@@ -17,14 +21,259 @@
+        Each time the iterator is called it will return an InvokeAIGeneratorOutput
+        object. Use like this:
+
+           outputs = txt2img.generate(prompt='banana sushi', iterations=5)
+           for result in outputs:
+               print(result.image, result.seed)
+
+        In the typical case of wanting to get just a single image, iterations
+        defaults to 1 and do:
+
+           output = next(txt2img.generate(prompt='banana sushi')
+
+        Pass None to get an infinite iterator.
+
+           outputs = txt2img.generate(prompt='banana sushi', iterations=None)
+           for o in outputs:
+               print(o.image, o.seed)
+        
+        '''
+        model_name = self.params.model_name or self.model_manager.current_model
+        model_info: dict = self.model_manager.get_model(model_name)
+        model:StableDiffusionGeneratorPipeline = model_info['model']
+        model_hash = model_info['hash']
+        scheduler: Scheduler = self.get_scheduler(
+            model=model,
+            scheduler_name=self.params.scheduler
+        )
+        uc, c, extra_conditioning_info = get_uc_and_c_and_ec(prompt,model=model)
+
+        def _wrap_results(image: Image, seed: int, **kwargs):
+            nonlocal results
+            results.append(output)
+
+        generator = self.load_generator(model, self._generator_name())
+        if self.params.variation_amount > 0:
+            generator.set_variation(self.params.seed,
+                                    self.params.variation_amount,
+                                    self.params.with_variations)
+
+        generator_args = dataclasses.asdict(self.params)
+        generator_args.update(keyword_args)
+
+        iteration_count = range(iterations) if iterations else itertools.count(start=0, step=1)
+        for i in iteration_count:
+            results = generator.generate(prompt,
+                                         conditioning=(uc, c, extra_conditioning_info),
+                                         sampler=scheduler,
+                                         **generator_args,
+                                         )
+            output = InvokeAIGeneratorOutput(
+                image=results[0][0],
+                seed=results[0][1],
+                model_name = model_name,
+                model_hash = model_hash,
+                params=generator_args,
+            )
+            if callback:
+                callback(output)
+            yield output
+            
+    @classmethod
+    def schedulers(self)->List[str]:
+        '''
+        Return list of all the schedulers that we currently handle.
+        '''
+        return list(self.scheduler_map.keys())
+
+    def load_generator(self, model: StableDiffusionGeneratorPipeline, class_name: str):
+        module_name = f'invokeai.backend.generator.{class_name.lower()}'
+        module = importlib.import_module(module_name)
+        constructor = getattr(module, class_name)
```

Comment:
This is the sort of pattern one might use if we're designing a framework for downstream developers to add arbitrary generators.

But if we're not adding them dynamically and there are only a modest number of them, it would be more static-analysis-friendly to import them normally here.

Reply:
Ok.

Reply:
Agreed here. Composition or inheritance would be an easier pattern for contributors to understand as well.

Reply:
This is removed and replaced with simpler code.